cluster-autoscaler/main.go

```diff
@@ -179,6 +181,7 @@ var (
 	cordonNodeBeforeTerminate          = flag.Bool("cordon-node-before-terminating", false, "Should CA cordon nodes before terminating during downscale process")
 	daemonSetEvictionForEmptyNodes     = flag.Bool("daemonset-eviction-for-empty-nodes", false, "DaemonSet pods will be gracefully terminated from empty nodes")
 	userAgent                          = flag.String("user-agent", "cluster-autoscaler", "User agent used for HTTP calls.")
+	nodeInfoProcessorName              = flag.String("node-info-processor", "", "node-info processor type. Available values: ["+strings.Join(nodeinfobuilder.GetAvailableNodeInfoProcessors(), ",")+"]")
```

Comment:
This is solving the problem of processor configuration half-way. 

My thinking is that processors are intended as a way to implement support for advanced use-cases (that are likely relevant to a very few users) without further complicating core. Historically the assumption was that they would mostly be implemented in forks, but there is a growing interest in adding them to the main repo. As the number of processors grows the reviews will increasingly be a problem (this is exactly what happened with cloudproviders).  

I think we ultimately want to solve it the same way we did for providers - create a directory and start delegating ownership of individual processors. I'd like to make sure any solution for managing processors is going towards this ultimate goal. Some thoughts on this:
 * We want a consistent way to manage all processors, not just NodeInfoProcessor.
 * We need a way to compile CA with only relevant processors, similarly to how it can be compiled with just a single provider. 
 * I'm not sure I like the idea of dynamic registration via init(). Having a static configuration seems like a good balance - any new processor would have to be initially reviewed by core maintainer (owner of static config) to verify the general idea, but following that initial review the processor would be managed by its owner. Once again, this is basically the same model we use for cloudproviders (cloudprovider/builder plays that role there). 

--- 

Solving the general problem of processors obviously goes far beyond the scope of this PR. If you want to volunteer to tackle it I'm happy to help, but I'm obviously not going to ask you to do it. However, I'd rather not include a partial solution that I don't think is going in the direction we want to ultimately go to. 
For now maybe just add a specific flag for this feature? Something like --add-pod-info-to-template-nodes and just use it in a simple if statement in main.go.

Reply:
> Historically the assumption was that they would mostly be implemented in forks, but there is a growing interest in adding them to the main repo. 

i did not realize this, thanks for the clarification.

> I think we ultimately want to solve it the same way we did for providers - create a directory and start delegating ownership of individual processors.

this makes a lot of sense to me. the unfortunate side effect is that this PR would slow down a little until we can agree on the subpoints, but i think it would be nice to have this for future node processors.

Reply:
Yeah, I don't think it's fair to block this PR until we figure out the details. What I think we should do is merge the processor part of this PR as-is (modulo review comments) and move it to this new directory once we set it up.

Reply:
Hi @MaciekPytel and @elmiko

Thanks for the detailed and constructive reviews.

I agree with you the problem of processor configuration is “halfway” solved, I am glad we are on the same page and I am happy to implement a more generic feature.

What do you think about the following approach?

> We want a consistent way to manage all processors, not just NodeInfoProcessor.

Would you accept extending the “factory/builder” idea to other Processor interfaces? If each Processor has a unique ID, each Process implementation can register a builder function that will be executed if the user wants to enabled it.

> I'm not sure I like the idea of dynamic registration via init().

I also agree with you that we should be able to compile the CA with only relevant processors, my approach here with the `func init()` was to invert the registration logic. With to this pattern, all the implementation files of my Processors are in a dedicated `package`. Then, if we add `go build tags` on each processor implementation, the user would be able to control which processor implementation they want to include in their build.

In the `main.go` file, the only reference to the processor package implementation would be:

```golang
import (
    //...
    _ "k8s.io/autoscaler/cluster-autoscaler/processors/nodeinfos/podtemplates"
)
```

and the flag will only contains options that was part of the build

```golang
flag.String("node-info-processor", "", "node-info processor type. Available values: ["+strings.Join(nodeinfobuilder.GetAvailableNodeInfoProcessors(), ",")+"]")
```

> We need a way to compile CA with only relevant processors, similarly to how it can be compiled with just a single provider.

That's interesting, I don't have a strong opinion and I am happy to use the same pattern as in the `cloudprovider/builder package`. If we still want to support the fact that we can remove some implementation at build time, we can define two build functions in 2 different files, with different build tags.

```golang
// +build podtemplateprocessor

package podtemplateprocessor

func BuildProcessor(opts *core.AutoscalerOptions) NodeInfoProcessor {
    // Real implementation
}
```

```golang
// +build !podtemplateprocessor

package podtemplateprocessor

func BuildProcessor(opts *core.AutoscalerOptions) NodeInfoProcessor {
    return nil
}
```

and then 

```golang
package nodeinfos

// Build returns a new nodeinfos.NodeInfoProcessor instance with the requested implementation
func BuildNodeInfoProcessor(opts *core.AutoscalerOptions) (nodeinfos.NodeInfoProcessor, error) {
    switch opts.NodeInfosProcessorName {
    case podtemplateprocessor.Name
        return podtemplateprocessor.BuildProcessor(opts)
    }
    return nil
}

```

I can create an issue to start a separate discussion about this topic. I am more than happy to work on it, basically build this framework to register and build processors, merge it and then rebase and rework this PR to implement our use case.

let my know what you think.

Reply:
> Would you accept extending the “factory/builder” idea to other Processor interfaces? If each Processor has a unique ID, each Process implementation can register a builder function that will be executed if the user wants to enabled it.

i don't have an objection, but i would like to clarify a little. when you say factory/builder, are you saying that the factory could build a node processor given the unique ID? (eg make me a Foo processor)

> I also agree with you that we should be able to compile the CA with only relevant processors, my approach here with the func init() was to invert the registration logic. ...

your logic generally makes sense to me, but i am also in the "not liking init()" camp. perhaps we could follow the metrics approach and just have a registration function that gets called once from main?

> That's interesting, I don't have a strong opinion and I am happy to use the same pattern as in the cloudprovider/builder package.  ...

i agree with @MaciekPytel about re-using the build tag pattern that we have for the cloud providers. that would be really nice for the node processors too.

thanks for taking the time to write this up and for driving it forward @CharlyF !